### PR TITLE
Emoji support

### PR DESCRIPTION
**Pull request by tocsoft**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/Fonts/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
Work to enable support of more font features... specifically emoji but has added benefits for other font features too.
* Add support for 32bit code points #82 #80
* Adds support for reading COLR and CPAL font tables for substituting glyphs if using a color renderer. 
* Fixes a line gap issue that caused glyphs to be rendered 1/2 a line gap to high #107 

_Follow on work in ImageSharp.Drawing required to update our text renderer to actually provide the ability to render the emoji glyphs_

### Examples 
_with a hacked together renderer_

### CODE REVIEW

File: src/SixLabors.Fonts/FontInstance.cs [hunk 1]
```diff
@@ -35,14 +39,20 @@ public class FontInstance : IFontInstance
         /// <param name="horizontalMetrics">The horizontal metrics.</param>
         /// <param name="head">The head.</param>
         /// <param name="kern">The kern.</param>
-        internal FontInstance(NameTable nameTable, CMapTable cmap, GlyphTable glyphs, OS2Table os2, HorizontalMetricsTable horizontalMetrics, HeadTable head, KerningTable kern)
+        /// <param name="colrTable">The COLR table</param>
+        /// <param name="cpalTable">The CPAL table</param>
+        internal FontInstance(NameTable nameTable, CMapTable cmap, GlyphTable glyphs, OS2Table os2, HorizontalMetricsTable horizontalMetrics, HeadTable head, KerningTable kern, ColrTable? colrTable, CpalTable? cpalTable)
```
Comment (JimBobSquarePants) [inline src/SixLabors.Fonts/FontInstance.cs:44]: Good thing this is internal. The params list is getting awfully long.

File: src/SixLabors.Fonts/GlyphType.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace SixLabors.Fonts
+{
+    /// <summary>
+    /// Represents the various version of a glyph records.
+    /// </summary>
+    public enum GlyphType
+    {
+        /// <summary>
+        /// This is a fall back glyph dure to missing code point
```
Comment (JimBobSquarePants) [inline src/SixLabors.Fonts/GlyphType.cs:12]: `due`?

File: src/SixLabors.Fonts/Tables/General/CMap/Format12SubTable.cs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                    glyphId = (ushort)((charAsInt - seg.StartCodePoint) + seg.StartGlyphId);
+                    return true;
+                }
+            }
+
+            glyphId = 0;
+            return false;
+        }
+
+        public static IEnumerable<Format12SubTable> Load(IEnumerable<EncodingRecord> encodings, BinaryReader reader)
```
Comment (JimBobSquarePants) [inline src/SixLabors.Fonts/Tables/General/CMap/Format12SubTable.cs:41]: I'm curious. Are there performance implications caused by using `IEnumerable<T>`?

File: src/SixLabors.Fonts/Tables/General/CMapTable.cs [hunk 4]
```diff
@@ -12,40 +12,40 @@ namespace SixLabors.Fonts.Tables.General
...
-                        {
-                            table = t;
-                        }
+                        tbls.Add(t);
                     }
                 }
             }
 
-            this.table = table;
+            this.platformTables = tbls.ToArray();
```
Comment (JimBobSquarePants) [inline src/SixLabors.Fonts/Tables/General/CMapTable.cs:34]: Is there a reason we have to create a new array here?

File: src/SixLabors.Fonts/Tables/General/ColrTable.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Linq;
+using SixLabors.Fonts.Tables.General.Colr;
+using SixLabors.Fonts.Tables.General.Glyphs;
+
+namespace SixLabors.Fonts.Tables.General
+{
+    [TableName(TableName)]
```
Comment (JimBobSquarePants) [inline src/SixLabors.Fonts/Tables/General/ColrTable.cs:11]: What are these attributes used for?

File: src/SixLabors.Fonts/Tables/General/CMap/Format12SubTable.cs (hunk 3 above)
Comment (tocsoft) [reply src/SixLabors.Fonts/Tables/General/CMap/Format12SubTable.cs:41]: can't imagine so this is only called during font loading (which should really only be being called once for the life of the running app) and is immediately added to a [List of tables](https://github.com/SixLabors/Fonts/blob/f05fd390be8b276d4cc982aef8ec6274a144d6fc/src/SixLabors.Fonts/Tables/General/CMapTable.cs#L106)

File: src/SixLabors.Fonts/Tables/General/CMapTable.cs (hunk 4 above)
Comment (tocsoft) [reply src/SixLabors.Fonts/Tables/General/CMapTable.cs:34]: everytime we lookup a glyph from a code point (hot path) we have to iterate over the  `platformTables` collectino and i'm under the impression that `foreach`ing an `Array` is quicker than a `List<T>`

File: src/SixLabors.Fonts/Tables/General/ColrTable.cs (hunk 5 above)
Comment (tocsoft) [reply src/SixLabors.Fonts/Tables/General/ColrTable.cs:11]: the are used in the `TableLoader` class [here](https://github.com/SixLabors/Fonts/blob/f05fd390be8b276d4cc982aef8ec6274a144d6fc/src/SixLabors.Fonts/Tables/TableLoader.cs#L22) to help link a table loader class with a static load method to a font table name so the loader can easily lookup the stream offsets required for accessing the table data.

File: src/SixLabors.Fonts/GlyphType.cs (hunk 2 above)
Comment (tocsoft) [reply src/SixLabors.Fonts/GlyphType.cs]: yup fat fingers all the way.. 😀 ..sorted.

File: src/SixLabors.Fonts/Tables/General/CMapTable.cs (hunk 4 above)
Comment (JimBobSquarePants) [reply src/SixLabors.Fonts/Tables/General/CMapTable.cs:34]: I'd have to benchmark to be sure. Fine for now anyway....

**Pull request merged**